File: adding_one/Ruby/adding_one.rb
p "Adding One With Ruby!"

# Adds one to an array of integers
def add_one array, indx=nil
	# Initialize indx to last index in array
	if indx == nil
		indx = array.length - 1
	end

	# Exit for reccursion(Add a digit and Insert one to the beginning)
	if indx < 0
		array.insert(0,1)
		return array
	end

	# Carry the one if digit is 9
	if array[indx] >= 9
		array[indx] = 0
		indx-=1
		return add_one array, indx
	

	# Else just increment the digit
	else
		array[indx] +=1
	end

	return array

end

p add_one [1,2,3,4] #1,235

p add_one [9,9,9] #1,000

p add_one [3,7,8,0,9,9] #378,100

File: adding_one/C++/adding_one.cpp
#include <iostream>
#include <vector>
// Adds one to an array of integers
std::vector<int> add_one(std::vector<int> v, int indx);


int main() {

	std::vector<int> v {1,2,5};
	int last_indx = v.size()-1; // Start from end of vector.
	add_one(v, last_indx);

}

std::vector<int> add_one(std::vector<int> v, int indx) {
	
	// Base code for Reccursive function (Insert a new digit, one, at the beginning of the array).
	if(indx < 0) {
		v.insert(v.begin(),1);
		return v;
	}

	// Carry the one if digit is 9
	if(v[indx] >= 9) {
		v[indx] = 0;

		return add_one(v, --indx);
	}
	// Or just increment
	else
		v[indx]++;
	return v;
}